Fix #760 by making administrator_login & administrator_password optional for connection details

niiku opened the pull request:
<!--
Thank you for helping to improve Official Azure Provider!

Please read through https://git.io/fj2m9 if this is your first time opening a
Official Azure Provider pull request. Find us in https://crossplane.slack.com 
if you need any help contributing.
-->

### Description of your changes
As described in #760 disabling password auth (only enabling azure ad) results in a panic as no credential details are available. Within this PR checks are introduced which verify the existence of username/password and only then add it to the credential details. 

As general "inspiration" I oriented the change on how it's done for MSSQL: https://github.com/crossplane-contrib/provider-upjet-azure/blob/main/config/sql/config.go#L39

<!--
Briefly describe what this pull request does. Be sure to direct your
reviewers' attention to anything that needs special consideration.

We love pull requests that resolve an open Official Azure Provider issue. If yours does, you
can uncomment the below line to indicate which issue your PR fixes, for example
"Fixes #500":

-->
Fixes #760

I have:

- [x] Read and followed Crossplane's [contribution process].
- [] Run `make reviewable` to ensure this PR is ready for review. <- I somehow didn't get that command running. But I doubt it's because of my code changes..
- [] Added `backport release-x.y` labels to auto-backport this PR if necessary. <- Not too sure. I need input from the maintainers if it makes sense to backport?

### How has this code been tested
I run the fixed provider within my cluster with the example MR as described in the bug report. With the applied changes, the panic is not appearing and the .status of the FlexibleServer MR is updated when creating with disabled password auth. 

A successful uptest run of the resource: https://github.com/crossplane-contrib/provider-upjet-azure/actions/runs/9483541667

<!--
Before reviewers can be confident in the correctness of this pull request, it
needs to tested and shown to be correct. Briefly describe the testing that has
already been done or which is planned for this change.
-->

[contribution process]: https://git.io/fj2m9

jeanduplessis commented:
/test-examples="examples/dbforpostgresql/v1beta2/flexibleserver.yaml"

turkenf commented:
Hey @niiku, many thanks for your effort and interest here, we appreciate it.

> I could be mistaken, but this does not seem to be related to my changes at all.

Yes, this is not related to the change you made, the e2e test of the current version of the resource seems to be unverified. Dependent resources need to be checked and selectors need to be adjusted.

niiku commented:
@turkenf Thanks a lot for your answer! I could have a look at the e2e tests. As far as I understood the code base, there aren't that many unit tests done for config.go - so I wanted to get some input how to best "bake in" the use case I added in form of tests. Might you guide me if you rather would like to see unit tests (as this might be well testable on this level) and/or have an e2e for this case as well?

niiku commented:
@jeanduplessis @turkenf I think i might have fixed the e2e as simply the privat dns name wasn't set correctly. Could you let them run again?

turkenf commented:
/test-examples="examples/dbforpostgresql/v1beta2/flexibleserver.yaml"

turkenf commented:
Hey @niiku, I have manually checked the resource's `examples/dbforpostgresql/v1beta2/flexibleserver.yaml` YAML file except for the changes you have made so far and I think the following changes are considerable:

- We have an `uptest.upbound.io/timeout` annotation that we use in e2e tests, we use this annotation when the creation and deletion of the resource takes over 20 minutes. When I look at the Terraform [documentation](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/postgresql_flexible_server#timeouts), I think it would be useful to add the annotation `uptest.upbound.io/timeout: "3600"`, this gives us 1 hour for testing.
As an example: https://github.com/crossplane-contrib/provider-upjet-azure/blob/0a511b8f5330fc8a8dec7dea0c76e2ef4e16ffcc/examples/dbforpostgresql/v1beta1/virtualnetworkrule.yaml#L10

- It would be logical to remove the `spec.forProvider.zone` field of the `FlexibleServer` resource since it is optional, so it will automatically be assigned to the available zone.

turkenf commented:
/test-examples="examples/dbforpostgresql/v1beta2/flexibleserver.yaml"

niiku commented:
I added the following changes:
* Added autoGeneratePassword because of
```
2024-06-12T07:56:29.1795017Z     logger.go:42: 07:56:29 | case/0-apply |       message: 'create failed: async create failed: failed to create the resource:
2024-06-12T07:56:29.1796616Z     logger.go:42: 07:56:29 | case/0-apply |         [***0 `administrator_password` is required when `create_mode` is `Default` and
2024-06-12T07:56:29.1798019Z     logger.go:42: 07:56:29 | case/0-apply |         `authentication.password_auth_enabled` is set to `true`  []***]'
```
* Removed the zone to allow better distribution
* Updated PostgreSQL version to 16 as it's latest


I would not recommend to increase the timeout - as this setup should be provisioned in like ~10min - so 30min as timeout is sufficient - if it takes longer it just delays the time until we get feedback from the e2e tests. 

Could you run the e2e tests again?

turkenf commented:
/test-examples="examples/dbforpostgresql/v1beta2/flexibleserver.yaml"

turkenf commented:
/test-examples="examples/dbforpostgresql/v1beta2/flexibleserver.yaml"

turkenf reviewed (approved):
Many thanks for your effort @niiku 🙌
